lib/puppet/util/windows/service.rb

```diff
@@ -254,15 +254,26 @@ class QUERY_SERVICE_CONFIGW < FFI::Struct
     # @param [:string] service_name name of the service to start
     def start(service_name)
       open_service(service_name, SC_MANAGER_CONNECT, SERVICE_START | SERVICE_QUERY_STATUS) do |service|
-        # don't attempt to fail here if the service isn't stopped because windows error codes
-        # are likely more informative than ours and a failed call to StartServiceW will produce
-        # those errors
-        wait_for_pending_transition(service, SERVICE_STOP_PENDING, SERVICE_STOPPED)
+        # A truthy value return from pending transision indicates the service is now in
+        # SERVICE_RUNNING, a falsey value indicates the service is in another state
+        if wait_on_pending(service, [SERVICE_START_PENDING, SERVICE_CONTINUE_PENDING], SERVICE_RUNNING)
```

Comment:
Any reason why we have this check? In Puppet, `SERVICE_START_PENDING`, `SERVICE_CONTINUE_PENDING`, and `SERVICE_RUNNING` will report an ensure of `running`, so we will not ever call this method. Same case with the `stop` method (`SERVICE_STOP_PENDING`, `SERVICE_STOPPED` will report ensure `stopped`, so `stop` won't be called either).

I think it would instead be useful to have these methods enforce the pre-condition of valid initial states. For example in start, we only want this called iff the service is in `STOP_PENDING` or `STOPPED`, otherwise we throw an exception. That's what the transition diagram says. Note that for `CONTINUE_PENDING`, we'd likely want a separate `resume` method in the service module.

Reply:
The idea with designing `start`, `stop` this way (and in the future, `pause`/`resume`) is to have these methods strictly adhere to the state diagram and represent actual state transitions. Depending on the commonality between these methods, we can possibly design a helper `transition_service_state(service, possible_initial_states, pending_state, final_state, &block)` where the `block` would perform the state transition (e.g. call `StartService` or `ControlService`). This method would look something like:
```
def transition_service_state(service, possible_initial_states, pending_state, final_state, &block)
  unless possible_initial_states.include?(service_state)
    raise error
  end

  // Perform the state transition
  block.call(service)
  unless wait_for_state_transition(service, [pending_state, final_state])
    raise Error
  end
  unless wait_for_pending_transition(service, pending_state, final_state)
    raise Error
  end
end
```

And we can implement start as:
```
def start(service)
  // This would pass-thru if the initial state is something other than SERVICE_STOP_PENDING
  wait_for_pending_transition(service, SERVICE_STOP_PENDING, SERVICE_STOPPED)
  transition_service_state(service, [SERVICE_STOPPED], SERVICE_START_PENDING, SERVICE_RUNNING) do |service|
    // call StartService here
  end
end
```

We can implement `stop`, `pause`, and `resume` in a similar fashion.

Reply:
I think you can actually go ahead and implement that helper in this PR. It would make it easier to review + set-it up to implement `pause` and `resume` in the future.